sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoRequiresStableInputTest.java

```diff
@@ -0,0 +1,166 @@
+ * ValidatesRunner test for the support of {@link
+ * org.apache.beam.sdk.transforms.DoFn.RequiresStableInput} annotation.
+ */
+@RunWith(JUnit4.class)
+public class ParDoRequiresStableInputTest {
+
+  private static final String VALUE = "value";
+  // SHA-1 hash of string "value"
+  private static final String VALUE_CHECKSUM = "f32b67c7e26342af42efabc674d441dca0a281c5";
+
+  private static class PairWithRandomKeyFn extends SimpleFunction<String, KV<String, String>> {
+    @Override
+    public KV<String, String> apply(String value) {
+      String key = UUID.randomUUID().toString();
+      return KV.of(key, value);
+    }
+  }
+
+  private static class MakeSideEffectAndThenFailFn extends DoFn<KV<String, String>, String> {
+    private final String outputPrefix;
+
+    private MakeSideEffectAndThenFailFn(String outputPrefix) {
+      this.outputPrefix = outputPrefix;
+    }
+
+    @RequiresStableInput
+    @ProcessElement
+    public void processElement(ProcessContext c) throws Exception {
+      MatchResult matchResult = FileSystems.match(outputPrefix + "*");
+      boolean firstTime = (matchResult.metadata().size() == 0);
+
+      KV<String, String> kv = c.element();
+      writeTextToFileSideEffect(kv.getValue(), outputPrefix + kv.getKey());
+      if (firstTime) {
+        throw new Exception("Deliberate failure: should happen only once.");
+      }
+    }
+
+    private static void writeTextToFileSideEffect(String text, String filename) throws IOException {
+      ResourceId rid = FileSystems.matchNewResource(filename, false);
+      WritableByteChannel chan = FileSystems.create(rid, "text/plain");
+      chan.write(ByteBuffer.wrap(text.getBytes(Charset.defaultCharset())));
+      chan.close();
+    }
+  }
+
+  private static void runRequiresStableInputPipeline(RequiresStableInputTestOptions options) {
+    Pipeline p = Pipeline.create(options);
+
+    PCollection<String> singleton = p.apply("CreatePCollectionOfOneValue", Create.of(VALUE));
+    singleton
+        .apply("Single-PairWithRandomKey", MapElements.via(new PairWithRandomKeyFn()))
+        .apply(
+            "Single-MakeSideEffectAndThenFail",
+            ParDo.of(new MakeSideEffectAndThenFailFn(options.getParDoSingleOutputPrefix())));
+    singleton
+        .apply("Multi-PairWithRandomKey", MapElements.via(new PairWithRandomKeyFn()))
+        .apply(
+            "Multi-MakeSideEffectAndThenFail",
+            ParDo.of(new MakeSideEffectAndThenFailFn(options.getParDoMultiOutputPrefix()))
+                .withOutputTags(new TupleTag<>(), TupleTagList.empty()));
+
+    p.run().waitUntilFinish();
+  }
+
+  @BeforeClass
+  public static void setup() {
+    PipelineOptionsFactory.register(TestPipelineOptions.class);
+  }
+
+  /**
+   * Option for ParDoRequiresStableInputTest.
+   *
+   * <p>Define the output prefixes for tests on {@link ParDo.SingleOutput} and {@link
+   * ParDo.MultiOutput}.
+   */
+  public interface RequiresStableInputTestOptions extends TestPipelineOptions {
```

Comment:
I believe we should treat this as an IT instead of a regular ValidatesRunner test because it relies on external state **OR** you can keep it as a ValidatesRunner test by having the test rely on AttemptedMetrics since it is the only thing that you can modify that is part of the Apache Beam model and also changes even if a bundle fails. Using this would allow you to avoid files (which is a hassle in testing (who cleans them up, what if the IO system is down, requires extra configuration from a user)) and then you could keep this test as a ValidatesRunner test which is quite valuable. It would require runners to support attempted metrics in addition to requires stable input to test the feature though.

Reply:
The second option sounds very good to me. I am looking into it. Thanks for your suggestion.

Reply:
I think this should be an integration test if it requires more access than the current ValidatesRunner tests.